// @sap/cds/bin/deploy/to-hana/cfUtil.js
const cp = require('child_process');
const fsp = require('fs').promises;
const os = require('os');
const path = require('path');

const cds = require('../../../lib');
const LOG = cds.log ? cds.log('deploy') : console;
const DEBUG = cds.debug('deploy');

const { bold } = require('../../utils/term');

const CF_COMMAND = 'cf';

const POLL_COUNTER = 40;
const POLL_DELAY = 2500; //ms

const OPERATION_STATE_INITIAL = 'initial';
const OPERATION_STATE_IN_PROGRESS = 'in progress';
const OPERATION_STATE_FAILED = 'failed';
const OPERATION_STATE_SUCCEEDED = 'succeeded';


class CfUtil {

    _clear() {
        this.spaceInfo = null;
    }

    async _sleep(ms) {
        return new Promise((resolve) => {
            setTimeout(resolve, ms);
        });
    }

    async _cfRun(...args) {
        const cmdLine = `${CF_COMMAND} ${args.join(' ')}`;
        if (DEBUG) {
            // eslint-disable-next-line no-console
            console.time(cmdLine);
            LOG.debug('>>> ' + cmdLine);
        }

        try {
            return await new Promise((resolve, reject) => {
                const child = cp.spawn(CF_COMMAND, args);

                let stdout = '';
                child.stdout.on('data', (data) => {
                    stdout += data;
                });

                let stderr = '';
                child.stderr.on('data', (data) => {
                    stderr += data;
                });

                child.on('error', (err) => {
                    DEBUG && LOG.debug(`${stdout}\n${stderr}`);
                    if (err.code === 'ENOENT') {
                        reject(new Error(`Command ${bold(CF_COMMAND)} not found. Make sure to install the Cloud Foundry Command Line Interface.`));
                    } else {
                        reject(err);
                    }
                });

                child.on('close', (code) => {
                    DEBUG && LOG.debug(`${stdout}\n${stderr}`);
                    if (!code) {
                        resolve(stdout.trim());
                    } else {
                        const errorMessage = `${stdout}\n${stderr}`;
                        reject(new Error(errorMessage.trim()));
                    }
                });
            });
        } finally {
            // eslint-disable-next-line no-console
            DEBUG && console.timeEnd(cmdLine);
        }
    }

    async _cfRequest(urlPath, queryObj, bodyObj) {
        if (queryObj) {
            const entries = Object.entries(queryObj);
            const queryStr = entries.map(([key, value]) => {
                // commas cause problems in cf curl when not double encoded
                value = value.replace(/,/g, encodeURIComponent(','));
                return `${key}=${encodeURIComponent(value)}`;
            }).join('&');

            urlPath = urlPath + `?${queryStr}`;
        }

        // cf curl PATH [-iv] [-X METHOD] [-H HEADER]... [-d DATA] [--output FILE]
        const args = ['curl', urlPath];
        if (bodyObj) {
            args.push('-d');
            args.push(JSON.stringify(bodyObj)); // cfRun uses spawn so no special handling for quotes on cli required
        }

        const response = await this._cfRun(...args);
        if (!response) {
            throw new Error(`The response from the server was empty. Maybe your token is expired. Run the command again and re-log on in case the problem persists.`);
        }

        const result = JSON.parse(response);
        if (result.errors) {
            const errorMessage = result.errors.map((entry) => `${entry.title}: ${entry.detail} (${entry.code})`).join('\n');
            throw new Error(errorMessage);
        }
        return result;
    }

    _extract(string, pattern, errorMsg) {
        const match = string.match(pattern);
        if (match?.[1]) {
            return match[1];
        }
        throw new Error(errorMsg);
    }

    async getCfTargetFromConfigFile() {
        const cfHome = process.env.CF_HOME || process.env.cf_home || path.join(os.homedir(), '.cf');
        try {
            const fileContent = await fsp.readFile(path.join(cfHome, 'config.json'));
            const config = JSON.parse(fileContent);
            if (config) {
                return {
                    apiEndpoint: this._extract(config.Target, /\s*(.+)\s*/, `CF API endpoint is missing. Use 'cf login' to login.`),
                    org: this._extract(config.OrganizationFields.Name, /\s*(.+)\s*/, `CF org is missing. Use 'cf target -o <ORG> to specify.`),
                    space: this._extract(config.SpaceFields.Name, /\s*(.+)\s*/, `CF space is missing. Use 'cf target -s <SPACE>' to specify.`)
                }
            }
        } catch (err) {
            LOG.debug(`getCfTargetFromConfigFile: ${err}`);
        }
    }

    async getCfTargetFromCli() {
        const result = await this._cfRun('target');
        return {
            apiEndpoint: this._extract(result, /api endpoint\s*:\s*([^\s]+)/i, `CF API endpoint is missing. Use 'cf login' to login.`),
            user: this._extract(result, /user\s*:\s*(.+)/i, `CF user is missing. Use 'cf login' to login.`),
            org: this._extract(result, /org\s*:\s*(.+)/i, `CF org is missing. Use 'cf target -o <ORG> to specify.`),
            space: this._extract(result, /space\s*:\s*(.+)/i, `CF space is missing. Use 'cf target -s <SPACE>' to specify.`),
        };
    }

    async getCfTarget() {
        // check if token is valid or expired / missing
        await this._cfRun('oauth-token');
        return await this.getCfTargetFromConfigFile() || await this.getCfTargetFromCli();
    }

    async getCfSpaceInfo() {
        if (!this.spaceInfo) {
            DEBUG && LOG.debug('getting space info');

            const target = await this.getCfTarget();

            const { org, space } = target;
            const orgs = await this._cfRequest(`/v3/organizations`, { names: org });
            if (!orgs?.resources?.length) {
                throw new Error(`CF org ${bold(org)} not found!`);
            }

            const orgGuid = orgs.resources[0].guid;
            const spaces = await this._cfRequest(`/v3/spaces`, { names: space, organization_guids: orgGuid });
            if (!spaces?.resources?.length) {
                throw new Error(`CF space ${bold(space)} not found in org ${bold(org)}!`);
            }

            const spaceGuid = spaces.resources[0].guid;

            this.spaceInfo = Object.assign({}, target, { orgGuid, spaceGuid });
        }

        return this.spaceInfo;
    }

    async getService(serviceName, showMessage = true) {
        showMessage && LOG.log(`Getting service ${bold(serviceName)}`);
        const spaceInfo = await this.getCfSpaceInfo();

        let counter = POLL_COUNTER;
        while (counter > 0) {
            counter--;
            const serviceInstances = await this._cfRequest('/v3/service_instances', {
                names: serviceName,
                space_guids: spaceInfo.spaceGuid,
                organization_guids: spaceInfo.orgGuid
            });
            if (!serviceInstances?.resources?.length) {
                return null;
            }

            const serviceInstance = serviceInstances.resources[0];
            switch (serviceInstance?.last_operation?.state?.toLowerCase()) {
                case OPERATION_STATE_INITIAL:
                case OPERATION_STATE_IN_PROGRESS:
                    await this._sleep(POLL_DELAY);
                    break;

                case OPERATION_STATE_SUCCEEDED:
                    return serviceInstance;

                case OPERATION_STATE_FAILED:
                    throw new Error(`The returned service reported state '${OPERATION_STATE_FAILED}'.\n${JSON.stringify(serviceInstance, null, 4)}`);

                default:
                    LOG.log(`Unsupported server response state '${serviceInstance?.last_operation?.state}'. Waiting for next response.`);
                    break;
            }
        }

        throw new Error(`Timeout occurred while getting service ${bold(serviceName)}`);
    }


    async getOrCreateService(serviceOfferingName, planName, serviceName, options) {

        const probeService = await this.getService(serviceName, false);
        if (probeService) {
            LOG.log(`Getting service ${bold(serviceName)}`);
            return probeService;
        }

        LOG.log(`Creating service ${bold(serviceName)} - please be patient...`);

        const spaceInfo = await this.getCfSpaceInfo();

        const servicePlan = await this._cfRequest(`/v3/service_plans`, {
            names: planName,
            space_guids: spaceInfo.spaceGuid,
            organization_guids: spaceInfo.orgGuid,
            service_offering_names: serviceOfferingName
        });

        if (!servicePlan?.resources?.length) {
            throw new Error(`No service plans found`);
        }

        const body = {
            type: 'managed',
            name: serviceName,
            tags: [serviceOfferingName],
            relationships: {
                space: {
                    data: {
                        guid: spaceInfo.spaceGuid
                    }
                },
                service_plan: {
                    data: {
                        guid: servicePlan.resources[0].guid
                    }
                }
            }
        }

        if (options) {
            body.parameters = { ...options };
        }

        const postResult = await this._cfRequest('/v3/service_instances', undefined, body);
        if (postResult.errors) {
            throw new Error(postResult.errors[0].detail);
        }

        const newService = await this.getService(serviceName, false);
        if (newService) {
            return newService;
        }

        throw new Error(`Could not create service ${bold(serviceName)}`);
    }


    async getServiceKey(serviceInstance, serviceKeyName, showMessage = true) {
        showMessage && LOG.log(`Getting service key ${bold(serviceKeyName)}`);

        let counter = POLL_COUNTER;
        while (counter > 0) {
            counter--;
            const bindings = await this._cfRequest(`/v3/service_credential_bindings`, { names: serviceKeyName, service_instance_guids: serviceInstance.guid });
            if (!bindings?.resources?.length) {
                return null;
            }

            const binding = bindings.resources[0];
            switch (binding?.last_operation?.state?.toLowerCase()) {
                case OPERATION_STATE_INITIAL:
                case OPERATION_STATE_IN_PROGRESS:
                    await this._sleep(POLL_DELAY);
                    break;

                case OPERATION_STATE_SUCCEEDED: {
                    const keyDetails = await this._cfRequest(`/v3/service_credential_bindings/${encodeURIComponent(binding.guid)}/details`);
                    return keyDetails.credentials;
                }

                case OPERATION_STATE_FAILED:
                    throw new Error(`The returned binding reported state '${OPERATION_STATE_FAILED}'.\n${JSON.stringify(binding, null, 4)}`);

                default:
                    LOG.log(`Unsupported server response state '${binding?.last_operation?.state}'. Waiting for next response.`);
                    break;
            }
        }

        throw new Error(`Timeout occurred while getting service key ${bold(serviceKeyName)}`);
    }


    async getOrCreateServiceKey(serviceInstance, serviceKeyName) {

        const serviceKey = await this.getServiceKey(serviceInstance, serviceKeyName, false);
        if (serviceKey) {
            LOG.log(`Getting service key ${bold(serviceKeyName)}`);
            return serviceKey;
        }

        LOG.log(`Creating service key ${bold(serviceKeyName)} - please be patient...`);

        const body = {
            type: 'key',
            name: serviceKeyName,
            relationships: {
                service_instance: {
                    data: {
                        guid: serviceInstance.guid
                    }
                }
            },
            parameters: {
                permissions: 'development'
            }
        }

        const postResult = await this._cfRequest('/v3/service_credential_bindings', undefined, body);
        if (postResult.errors) {
            throw new Error(postResult.errors[0].detail);
        }

        const newServiceKey = await this.getServiceKey(serviceInstance, serviceKeyName, false);
        if (newServiceKey) {
            return newServiceKey;
        }

        throw new Error(`Could not create service key ${bold(serviceKeyName)}`);
    }
}

module.exports = new CfUtil();
